### ISSUE TITLE
Please add getParameterIndex in ITestResult interface

### ISSUE THREAD

**Issue by ghenadied** [cbeust/testng#1671]
### TestNG Version
latest
> Note: only the latest version is supported

### Expected behavior
Possibility to grab in reports parameter index (from data provider) of a test method.

### Actual behavior
There is no way one could extract in custom reporter (nor in default one) the parameter index test method was run with. Since now there are methods added in **TestResult** class to set and get parameterindex, but get is used nowhere.
It's pretty useful when one has a method iterated with data from data provider and there are a lot of entries in data provider with slight changes each and a lot of values for a single iteration, it's hard to find the failed tests by looking in report and data provider data entries so that they can be correlated. Even more, it's much easier to investigate when method is run with parallel data provider.

Sample test case: having a method which iterates over hundred times with some (5 maps), each map having around 10 fields, each iteration having only one or two field changed.

### Is the issue reproductible on runner?
all

### Test case sample
> Please, share the test case (as small as possible) which shows the issue

**Comment by krmahadevan** [comment]
@ghenadied - Adding a new method to `ITestResult` interface will break backward compatibility for others when they upgrade to the new version. Once TestNG moves to JDK8 this can be added up as a default method to the `ITestResult` interface without any such breakage.

In the meantime, can you not do something like this ?

```java
import org.testng.IReporter;
import org.testng.ISuite;
import org.testng.ISuiteResult;
import org.testng.ITestContext;
import org.testng.ITestResult;
import org.testng.annotations.DataProvider;
import org.testng.annotations.Listeners;
import org.testng.annotations.Test;
import org.testng.internal.TestResult;
import org.testng.xml.XmlSuite;

import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.List;

@Listeners(SampleTestClass.ExampleReporter.class)
public class SampleTestClass {

    @Test(dataProvider = "dp")
    public void anotherTest(String i) {
        System.err.println(i);
    }

    @DataProvider(name = "dp")
    public Object[][] getData(Method method) {
        return new Object[][]{
                {"TestNG"},
                {"Selenium"}
        };
    }

    public static class ExampleReporter implements IReporter {

        @Override
        public void generateReport(List<XmlSuite> xmlSuites, List<ISuite> suites, String outputDirectory) {
            suites.forEach(
                    iSuite -> iSuite.getResults().values()
                            .forEach(ExampleReporter::printSuiteResults)
            );
        }

        private static void printSuiteResults(ISuiteResult iSuiteResult) {
            printTestResults(iSuiteResult.getTestContext());
        }

        private static void printTestResults(ITestContext context) {
            context.getPassedTests().getAllResults().forEach(ExampleReporter::printIndividualTestMethodDetails);
        }

        private static void printIndividualTestMethodDetails(ITestResult testResult) {
            if (testResult instanceof TestResult) {
                int index = ((TestResult) testResult).getParameterIndex();
                String params = Arrays.toString(testResult.getParameters());
                System.err.println("Index = " + index + " values = " + params);
            }
        }
    }
}
```

Here's the output

```
TestNG
Selenium

===============================================
Default Suite
Total tests run: 2, Failures: 0, Skips: 0
===============================================

Index = 0 values = [TestNG]
Index = 1 values = [Selenium]

Process finished with exit code 0

```

**Comment by ghenadied** [comment]
Hi  @krmahadevan ,

Understood. Yes, this is the way to go then. 
I'll wait for the issues with parallel modes to be solved on current branch then to make use of it since getParameterIndex wasn't added until recently in TestResult class (6.13.1 guess).
Still, I'll put a sticker note to not forget once upgrading TestNG to Java8 :)

Thanks a lot!

**Comment by krmahadevan** [comment]
@ghenadied 

> since getParameterIndex wasn't added until recently in TestResult class (6.13.1 guess).

This was introduced in **TestNG 6.12** as part of #1420. So you can work with **TestNG 6.12** (or) even **TestNG 6.13.1** (the latest released version as of today)

> I'll wait for the issues with parallel modes to be solved on current branch

what issues are you talking about ?  The only one that I am aware of is #1636 and its been fixed in `6.14-SNAPSHOT`. Also there's a [simple work-around](https://github.com/cbeust/testng/issues/1636#issuecomment-349182219) which should help you get past this
